pkg/controllers/clusterresourceplacement/placement_controller.go

```diff
@@ -333,3 +339,116 @@ func (r *Reconciler) updatePlacementAppliedCondition(placement *fleetv1alpha1.Cl
+		latestPolicySnapshotIndex++
+		latestPolicySnapshot = &fleetv1.ClusterPolicySnapshot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf(fleetv1.PolicySnapshotNameFmt, crp.Name, latestPolicySnapshotIndex),
+				Labels: map[string]string{
+					fleetv1.CRPTrackingLabel:      crp.Name,
+					fleetv1.IsLatestSnapshotLabel: strconv.FormatBool(true),
+					fleetv1.PolicyIndexLabel:      strconv.Itoa(latestPolicySnapshotIndex),
+				},
+				OwnerReferences: []metav1.OwnerReference{ownerReference(crp)},
+			},
+			Spec: fleetv1.PolicySnapShotSpec{
+				Policy:     crp.Spec.Policy,
+				PolicyHash: []byte(policyHash),
+			},
+		}
+		if err := r.Client.Create(ctx, latestPolicySnapshot); err != nil {
+			klog.ErrorS(err, "Failed to create new clusterPolicySnapshot", "clusterPolicySnapshot", klog.KObj(latestPolicySnapshot))
+			return ctrl.Result{}, err
+		}
+	}
+	// create clusterResourceSnapshot
+	// update the status based on the latestPolicySnapshot status
+	// update the status based on the work
+	return ctrl.Result{}, nil
+}
+
+func ownerReference(crp *fleetv1.ClusterResourcePlacement) metav1.OwnerReference {
+	return metav1.OwnerReference{
+		APIVersion:         crp.GroupVersionKind().GroupVersion().String(),
+		Kind:               crp.GroupVersionKind().Kind,
+		Name:               crp.GetName(),
+		UID:                crp.GetUID(),
+		BlockOwnerDeletion: pointer.Bool(true),
+		Controller:         pointer.Bool(true),
+	}
+}
+
+// lookupLatestClusterPolicySnapshot finds the index of the latest cluster policy snapshot in the array and its policyIndex.
+// There will be only one active policy snapshot.
+// If there are no active policy snapshots, find the one whose policyIndex label is the largest.
+// Return error when the policyIndex label is invalid.
+func lookupLatestClusterPolicySnapshot(list *fleetv1.ClusterPolicySnapShotList) (*fleetv1.ClusterPolicySnapshot, int, error) {
+	if len(list.Items) == 0 {
+		return nil, -1, nil
+	}
+	index := -1           // the index of the cluster policy snapshot array
+	lastPolicyIndex := -1 // the assigned policy index of the cluster policy snapshot
+	for i, snapshot := range list.Items {
+		policyIndex, err := parsePolicyIndexFromLabel(&list.Items[i])
+		if err != nil {
+			return nil, -1, err
+		}
+		if v := snapshot.Labels[fleetv1.IsLatestSnapshotLabel]; v == strconv.FormatBool(true) {
+			return &list.Items[i], policyIndex, nil
+		}
+		if lastPolicyIndex < policyIndex {
+			index = i
+			lastPolicyIndex = policyIndex
+		}
+	}
+	return &list.Items[index], lastPolicyIndex, nil
+}
+
+func parsePolicyIndexFromLabel(s *fleetv1.ClusterPolicySnapshot) (int, error) {
+	indexLabel := s.Labels[fleetv1.PolicyIndexLabel]
+	v, err := strconv.Atoi(indexLabel)
+	if err != nil {
+		klog.ErrorS(errors.New("invalid clusterPolicySnapshot policyIndex label"), "Failed to parse the policy index label", "clusterPolicySnapshot", klog.KObj(s), "policyIndexLabel", indexLabel)
+		// should never happen
```

Comment:
Assuming we will add the deny policy to reject customer's modification and create alerts for this kind of unexpected errors, our controller/system should be in the wrong state when hitting this issue.

The simple way is to rollback our controllers instead of swallowing the error.

I was thinking skipping the snapshots with invalid labels before but figured out it did not bring much benefits compared with the its complexity.

WDYT?

Reply:
yeah, those cases should not happen especially when we will have webhooks in place. But the question is still what should do about those? Here are some options

- Don't check those "should never happen" case in the code at all. Let the system crash which will trigger alerts.
- Check and raise alerts in the code directly (not sure if there is a way)
- Check and add an error metrics and alert based on those metrics

I am just not sure that returning an error is a good way to handle this type of "should never happen" error since it will stuck and the system can't recover on its own.

Reply:
agree with you. When we hit these cases, oncall should be involved.
Prefer to take option3 and will add the metrics in a separate PR.

option 1 will impact the whole service and the error may be only caused by one CR.
The other CRs could work correctly.